### PR TITLE
Set pages allocated in the PE loader to NX after allocation and NX+RO after the image is loaded into the memory

### PR DESCRIPTION
**Pull request by TaylorBeebe**

### CODE REVIEW

File: MdeModulePkg/Library/DxeSmmMemoryProtectionHobLib/CommonMemoryProtectionHobLib.c [hunk 1]
```diff
@@ -132,19 +132,27 @@ HeapGuardPolicy.SmmPageGuard are active.\n",
...
+// DxeNxProtectionPolicy.EfiBootServicesCode, \
+// and DxeNxProtectionPolicy.EfiRuntimeServicesCode \
+// must be set to ZERO. Setting all to ZERO \
+// in the memory protection settings global.\n",
+//       __FUNCTION__
+//       ));
+        DEBUG ((
       DEBUG_WARN,
-      "%a: - DxeNxProtectionPolicy.EfiLoaderCode, \
-DxeNxProtectionPolicy.EfiBootServicesCode, \
```
Comment (Bethk85) [inline MdeModulePkg/Library/DxeSmmMemoryProtectionHobLib/CommonMemoryProtectionHobLib.c]: EfiBootServicesCode

Comment (Bethk85) [inline MdeModulePkg/Library/DxeSmmMemoryProtectionHobLib/CommonMemoryProtectionHobLib.c]: MdeModulePkg/Library/DxeSmmMemoryProtectionHobLib/CommonMemoryProtectionHobLib.c

**Review by Bethk85 — APPROVED** [review]

**Review by Bethk85 — APPROVED** [review]

**Review by Bethk85 — APPROVED** [review]

**Review by Bethk85 — APPROVED** [review]

File: MdeModulePkg/Core/Dxe/Image/Image.c [hunk 2]
```diff
@@ -624,6 +628,15 @@ CoreLoadPeImage (
     return EFI_UNSUPPORTED;
   }
 
+  // MU_CHANGE START
+  if (!(Image->ImageContext.SupportsNx) && Image->ImageContext.ImageCodeMemoryType == EfiLoaderCode) {
+    if (!gMPS.ImageProtectionPolicy.Fields.AllowImagesWithoutNxFlag) {
+      return EFI_INVALID_PARAMETER;
```
Comment (spbrogan) [inline MdeModulePkg/Core/Dxe/Image/Image.c]: not sure about EFI_INVALID_PARAMETER. Usually this is reserved for when you are calling a function with invalid parameters.  In this case your policy won't let you load this image but you (the developer/consumer of API) called the function correctly.  

We have return values like security violation, etc.  Should look to see what else we can use.

File: MdeModulePkg/Core/Dxe/Image/Image.c [hunk 3]
```diff
@@ -624,6 +628,15 @@ CoreLoadPeImage (
     return EFI_UNSUPPORTED;
   }
 
+  // MU_CHANGE START
+  if (!(Image->ImageContext.SupportsNx) && Image->ImageContext.ImageCodeMemoryType == EfiLoaderCode) {
```
Comment (spbrogan) [inline MdeModulePkg/Core/Dxe/Image/Image.c:637]: I would appreciate seeing more code comments.  
With UEFI code which is shared by large community, integrated by dozens of partners, and lives on for years I think adding comments to anything important and/or anything with even a remotely complex conditional or critical conditional is really helpful in long term maintainability and quick understanding.

**Review by spbrogan — APPROVED** [review]

File: MdeModulePkg/Core/Dxe/Image/Image.c (hunk 3 above)
Comment (makubacki) [inline MdeModulePkg/Core/Dxe/Image/Image.c:637]: Perhaps move this code to a function in MemoryProtection.c to keep cohesion between memory protection related functionality.

File: MdeModulePkg/Core/Dxe/Image/Image.c [hunk 4]
```diff
@@ -753,7 +766,36 @@ CoreLoadPeImage (
...
+            (VOID **) &mMemoryAttribute
+            );
+  }
 
+  if (mMemoryAttribute != NULL) {
+
+    Status = mMemoryAttribute->GetMemoryAttributes (
+                                mMemoryAttribute,
+                                Image->ImageContext.ImageAddress,
+                                EFI_PAGES_TO_SIZE (Image->NumberOfPages),
```
Comment (kuqin12) [inline MdeModulePkg/Core/Dxe/Image/Image.c]: Out of curiosity, does this region cover only the code? Or both the code and data?

Comment (TaylorBeebe) [reply MdeModulePkg/Core/Dxe/Image/Image.c]: Both code and data. Later, when ProtectUefiImageMu() is called, permissions are updated appropriately.

File: MdeModulePkg/Core/Dxe/Image/Image.c (hunk 3 above)
Comment (TaylorBeebe) [reply MdeModulePkg/Core/Dxe/Image/Image.c:637]: Moved to memoryprotection.c

File: MdeModulePkg/Core/Dxe/Image/Image.c (hunk 2 above)
Comment (TaylorBeebe) [reply MdeModulePkg/Core/Dxe/Image/Image.c]: Switched to EFI_SECURITY_VIOLATION

**Pull request merged**